packages/eslint-plugin/docs/rules/member-ordering.md

```diff
@@ -1,24 +1,46 @@
 # Require a consistent member declaration order (member-ordering)
 
-A consistent ordering of fields, methods and constructors can make interfaces, type literals, classes and class
-expressions easier to read, navigate and edit.
+A consistent ordering of fields, methods and constructors can make interfaces, type literals, classes and class expressions easier to read, navigate and edit.
 
 ## Rule Details
 
-This rule aims to standardize the way class declarations, class expressions, interfaces and type literals are structured.
+This rule aims to standardize the way class declarations, class expressions, interfaces and type literals are structured and ordered.
 
-It allows to group members by their type (e.g. `public-static-field`, `protected-static-field`, `private-static-field`, `public-instance-field`, ...). By default, their order is the same inside `classes`, `classExpressions`, `interfaces` and `typeLiterals` (note: not all member types apply to `interfaces` and `typeLiterals`). It is possible to define the order for any of those individually or to change the default order for all of them by setting the `default` option.
+### Grouping and sorting member groups
+
+It allows to group members by their type (e.g. `public-static-field`, `protected-static-field`, `private-static-field`, `public-instance-field`, ...) and enforce a certain order for these groups. By default, their order is the same inside `classes`, `classExpressions`, `interfaces` and `typeLiterals` (note: not all member types apply to `interfaces` and `typeLiterals`). It is possible to define the order for any of those individually or to change the default order for all of them by setting the `default` option.
+
+### Sorting members
+
+Besides grouping the members and sorting their groups, this rule also allows to sort the members themselves. You have 2 options: Sort all of them while ignoring their type or sort them inside of the member types (e.g. sort all fields in an interface alphabetically).
 
 ## Options
 
+These options allow to specify how to group the members and sort their groups.
+
 ```ts
 {
-  default?: Array<MemberType> | never
-  classes?: Array<MemberType> | never
-  classExpressions?: Array<MemberType> | never
+  default?: Array<MemberType> | 'never'
```

Comment:
I would consider reversing the options here (because the object form is preferred), and extracting duplicated types to make it look less scary.

> These options allow to specify how to group the members and sort their groups.
> 
> ```ts
> // for the allowed values of MemberTypes - see documentation below
> 
> type TypeOptions<T> =
>   // control the grouping and ordering within groups
>   | {
>     // The ordering of groups of members. Omit this to 
>     memberTypes?: Array<T> | 'never',
>     // Configure the sort order within each of the groups
>     order?: 'alphabetically' | 'as-written',
>   }
>   // ignore grouping and just sort alphabetically
>   | {
>     order: 'alphabetically',
>   };
> 
> type Options = {
>   default?: TypeOptions<MemberTypes> | 'never',
>   classes?: TypeOptions<MemberTypes> | 'never',
>   classExpressions?: TypeOptions<MemberTypes> | 'never',
>   interfaces?: TypeOptions<'field' | 'method' | 'constructor'> | 'never',
>   typeLiterals?: TypeOptions<'field' | 'method' | 'constructor'> | 'never',
> }
> ```

Reply:
I like this idea! However, the proposed API would be a breaking change compared to the current approach (`default: Array<MemberType> | never`) if I'm not mistaken.

There are 3 cases:
1. Sort groups, no member order enforced
2. Sort members, no group order enforced
3. Sort members within groups

To avoid a breaking change, we would probably have to use an API like this:
```
type TypeOptions<T> =
  Array<T> // Case 1
  | { // Case 2
    order: 'alphabetically' | 'as-written', // 'as-written' is kinda useless but maybe more consistent with case 3?
  }
  | { // Case 3
    memberTypes: Array<T>,
    order: 'alphabetically' | 'as-written',
  };
```

Or in a shorter version:
```
type TypeOptions<T> = Array<T> | {
  memberTypes?: Array<T>,
  order: 'alphabetically' | 'as-written',
};
```

If a breaking change is acceptable, I would prefer your proposal with case 1 being:
```
{
  memberTypes: Array<T>
}
```

Reply:
Hm... this is an interesting approach. I'm not sure if the documentation should diverge from the implementation. Maybe we deprecate the "old" syntax and remove it on the next major update?

Reply:
Hm... this is an interesting approach. I'm not sure I like the idea that documentation and implementation diverge. Maybe we could deprecate the "old" syntax and remove it with the next major update?

Reply:
@bradzacher I've included the new recommended syntax in the file. I'll try to resolve the other issues commit by commit - we should probably squash the commits afterwards